hydrolib/core/io/structure/models.py

```diff
@@ -0,0 +1,129 @@
+class Weir(Structure):
+    class Comments(Structure.Comments):
+        structure_type: Optional[str] = Field(
+            "Structure type; must read weir", alias="type"
+        )
+        allowed_flow_direction: Optional[str] = Field(
+            "Possible values: both, positive, negative, none.", alias="allowedFlowdir"
+        )
+
+        crest_level: Optional[str] = Field(
+            "Crest level of weir (m AD).", alias="crestLevel"
+        )
+        crest_width: Optional[str] = Field("Width of the weir (m).", alias="crestWidth")
+        correction_coefficient: Optional[str] = Field(
+            "Correction coefficient (-).", alias="corrCoeff"
+        )
+        use_velocity_height: Optional[str] = Field(
+            "Flag indicating whether the velocity height is to be calculated or not.",
+            alias="useVelocityHeight",
+        )
+
+    comments: Comments = Field(default_factory=Comments)
+
+    structure_type: Literal["weir"] = Field("weir", alias="type")
+    allowed_flow_direction: FlowDirection = Field(alias="allowedFlowdir")
+
+    crest_level: Union[float, Path] = Field(alias="crestLevel")
+    crest_width: Optional[float] = Field(None, alias="crestWidth")
+    correction_coefficient: float = Field(1.0, alias="corrCoeff")
+    use_velocity_height: bool = Field(True, alias="useVelocityHeight")
+
+
+class UniversalWeir(Structure):
+    class Comments(Structure.Comments):
+        structure_type: Optional[str] = Field(
+            "Structure type; must read universalWeir", alias="type"
+        )
+        allowed_flow_direction: Optional[str] = Field(
+            "Possible values: both, positive, negative, none.", alias="allowedFlowdir"
+        )
+
+        number_of_levels: Optional[str] = Field(
+            "Number of yz-Values.", alias="numLevels"
+        )
+        y_values: Optional[str] = Field(
+            "y-values of the cross section (m). (number of values = numLevels)",
+            alias="yValues",
+        )
+        z_values: Optional[str] = Field(
+            "z-values of the cross section (m). (number of values = numLevels)",
+            alias="zValues",
+        )
+        crest_level: Optional[str] = Field(
+            "Crest level of weir (m AD).", alias="crestLevel"
+        )
+        discharge_coefficient: Optional[str] = Field(
+            "Discharge coefficient c_e (-).", alias="dischargeCoeff"
+        )
+
+    comments: Comments = Field(default_factory=Comments)
+
+    structure_type: Literal["universalWeir"] = Field("universalWeir", alias="type")
+    allowed_flow_direction: FlowDirection = Field(alias="allowedFlowdir")
+
+    number_of_levels: int = Field(alias="numLevels")
+    y_values: List[float] = Field(alias="yValues")
+    z_values: List[float] = Field(alias="zValues")
+    crest_level: float = Field(alias="crestLevel")
+    discharge_coefficient: float = Field(alias="dischargeCoeff")
+
+    @validator("y_values", "z_values", pre=True)
+    def split_value_lists(cls, v):
```

Comment:
You wanted to re-use right? Check https://pydantic-docs.helpmanual.io/usage/validators/#reuse-validators. Or if you think that's ugly, at least split it to a utils and call it from here.